### ISSUE TITLE
Can't pull from public registry if logged in with a private registry

### ISSUE THREAD

**Issue by manuelmazzuola** [bflad/chef-docker#379]
We run in to the same stacktrace of issue #356 

```
[staging-test] out:   * docker_image[progrium/consul] action pull
 ================================================================================
[staging-test] out:     Error executing action `pull` on resource 'docker_image[progrium/consul]'
[staging-test] out:     ================================================================================
[staging-test] out:     
[staging-test] out:     Docker::Error::ArgumentError
[staging-test] out:     ----------------------------
[staging-test] out:     Must have id, got: {"id"=>nil, :headers=>{"X-Registry-Auth"=>"RegistryAuthHash198jsa821qidwqjidq1jowdojiq}}
[staging-test] out:     
[staging-test] out:     Cookbook Trace:
[staging-test] out:     ---------------
[staging-test] out:     /home/ubuntu/chef/manuel_chef/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/base.
rb:16:in `initialize'
[staging-test] out:     /home/ubuntu/chef/manuel_chef/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/image
.rb:119:in `new'
[staging-test] out:     /home/ubuntu/chef/manuel_chef/cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/image
.rb:119:in `create'
[staging-test] out:     /home/ubuntu/chef/manuel_chef/cookbooks/docker/libraries/provider_docker_image.rb:67:in `pull_image'
[staging-test] out:     /home/ubuntu/chef/manuel_chef/cookbooks/docker/libraries/provider_docker_image.rb:126:in `block in <class:D
ockerImage>'

```

We investigated adding a puts inside the `cookbooks/docker/files/default/vendor/gems/docker-api-1.22.2/lib/docker/image.rb` file.

```
def create(opts = {}, creds = nil, conn = Docker.connection, &block)
    credentials = creds.nil? ? Docker.creds : creds.to_json
    puts credentials
    headers = credentials && Docker::Util.build_auth_header(credentials) || {}
....
....
json = Docker::Util.fix_json(body)
puts json
image = json.reverse_each.find { |el| el && el.key?('id') }
```

and we got 

```
{"serveraddress":"https://docker.cloudesire.com/","user[1492/1821]
stry","password":"our_password","email":"dev@cloudesire.com"}
[staging-test] out: {"status"=>"Pulling repository docker.io/progrium/consul"}
[staging-test] out: {"errorDetail"=>{"message"=>"Authentication is required."}, "error"=>"Authentication is required."}
```

then we tried to reset the authentication header that the library send to the local daemon

```
    # Create a new Image.
    def create(opts = {}, creds = nil, conn = Docker.connection, &block)
      credentials = creds.nil? ? Docker.creds : creds.to_json
      puts credentials
      headers = credentials && Docker::Util.build_auth_header(credentials) || {}
      ####################################################################
      headers = {} 
      ####################################################################
      body = ''
      conn.post(
        '/images/create',
        opts,
        :headers => headers,
        :response_block => response_block(body, &block)
      )
      json = Docker::Util.fix_json(body)
      puts json
      image = json.reverse_each.find { |el| el && el.key?('id') }
      new(conn, 'id' => image && image.fetch('id'), :headers => headers)
    end

```

and **it works** !

So it is incorrectly using our private registry credentials to authenticate to the local docker daemon?

**Comment by someara** [comment]
That definitely seems to be the case =\

**Comment by someara** [comment]
I'm thinking the thing to do here may be to get rid of the `docker_registry` resource all together and handle authentication at the resource level. Thoughts?

**Comment by chasebolt** [comment]
i agree with handling auth at the resource level.

**Comment by manuelmazzuola** [comment]
The *pull* action fails because it tries to *post* the image `progrium/consul` to the local docker daemon passing to it our docker distribution credentials.
So it successfully pulled the image from the registry but fails to post it to the local docker daemon.

Why it pass some credentials to the local docker daemon ?

**Comment by someara** [comment]
I've patched the vendored docker-api gem and this seems to do the trick... mind sending a patch upstream?

**Comment by someara** [comment]
released as v1.0.21
